extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts

```diff
@@ -26,10 +26,38 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
 
 	private _watcher: vscode.FileSystemWatcher | undefined;
 
-	async getAllMarkdownDocuments() {
-		const resources = await vscode.workspace.findFiles('**/*.md', '**/node_modules/**');
-		const docs = await Promise.all(resources.map(doc => this.getMarkdownDocument(doc)));
-		return docs.filter(doc => !!doc) as SkinnyTextDocument[];
+	/**
+	 * Reads and parses all .md documents in the workspace.
+	 * Files are processed in batches, to keep the number of open files small.
+	 *
+	 * @returns Array of processed .md files.
+	 */
+	async getAllMarkdownDocuments(): Promise<SkinnyTextDocument[]> {
+		const maxConcurrent = 20;
+		const ignoreList: (string | vscode.RelativePattern)[] = ['**/node_modules/**'];
+		const docList: SkinnyTextDocument[] = [];
+
+		let isDone = false;
+		while (!isDone) {
+			await new Promise(async (resolve) => {
+				const resources = await vscode.workspace.findFiles('**/*.md', `{${ignoreList.join(',')}}`, maxConcurrent);
```

Comment:
Not sure about using the ignore list this way since it will get very large for large workspaces. For the batching, how about executing `findFiles` once and then batching the reads themselves?

Reply:
Just to clarify, the issue isn't that we are calling `workspace.findFiles('**/*.md')`, it is that we are taking that result and calling `getMarkdownDocument` for all `.md` files in the workspace at a single time.
Is that right?
If so, then I agree with you that this should be changed and an ignore list shouldn't be used like this.  I thought that we should be avoiding calling `findFiles` for all `.md` files at once and this was the only way to do it.

Reply:
Yes `findFiles` is safe. The problem is that we then try to open every found file in parallel (which could be many thousands of files)

Reply:
> Just to clarify, the issue isn't that we are calling `workspace.findFiles('**/*.md')`, it is that we are taking that result and calling `getMarkdownDocument` for all `.md` files in the workspace at a single time.
> 
> Is that right?
> 
> If so, then I agree with you that this should be changed and an ignore list shouldn't be used like this.  I thought that we should be avoiding calling `findFiles` for all `.md` files at once and this was the only way to do it.

> ###- [ ] - - - - - - - [x] - - - - - - - - - - - - - Uat zos lob ndrangt ndangt ndrangt @nrayburn-tech [_![image](https://user-images.githubusercontent.com/84305127/119748532-aa4eea80-```

```bec7-11eb-8e6e-6fa2deea02ae.png)_]()